### PR TITLE
Add InstallPlan invariant assertion checking, revealing cyclic dep problem

### PR DESCRIPTION
**Pull request by dcoutts**
It turns out that the install plan elaboration is constructing cyclic plans in some cases. The effect is that executing the plan simply misses out anything that depends on the packages involved in the cycle. This is probably the cause of #3996

With this patch such cases will fail with an assertion such as:

    internal error in InstallPlan.fromSolverInstallPlanWithProgress:
    The following packages are involved in a dependency cycle
    hspec-discover-2.3.1-da63d0b4e952e7949a113646e4af0aac925a4d864a1db650..

The cause is clearly in the caller of `fromSolverInstallPlanWithProgress` and the only caller of that is `ProjectPlanning.elaborateInstallPlan`. The problem appears to be to do with the intra-package dependencies. The hspec-discover-2.3.1 example is one where it has a lib and an exe and the exe depends on the lib. The resulting elaborated plan ends up with the lib component with a self-dependency which is clearly wrong.

@ezyang would you mind having a look?

### CODE REVIEW

**Comment by ezyang** [comment]
Travis failure is due to DNS outage.

File: cabal-install/Distribution/Client/InstallPlan.hs [hunk 1]
```diff
@@ -223,17 +223,21 @@ type InstallPlan = GenericInstallPlan
 
 -- | Smart constructor that deals with caching the 'Graph' representation.
 --
-mkInstallPlan :: Graph (GenericPlanPackage ipkg srcpkg)
+mkInstallPlan :: (IsUnit ipkg, IsUnit srcpkg)
```
Comment (ezyang) [inline cabal-install/Distribution/Client/InstallPlan.hs]: We can get more accurate location tracing by requiring a CallStack (WithCallStack)

**Comment by ezyang** [comment]
LGTM.

File: cabal-install/Distribution/Client/InstallPlan.hs (hunk 1 above)
Comment (dcoutts) [inline cabal-install/Distribution/Client/InstallPlan.hs]: I thought of that but couldn't be bothered with the CPP required given the call stack stuff only works with newer ghcs.

Comment (ezyang) [inline cabal-install/Distribution/Client/InstallPlan.hs]: That's what Distribution.Compat.Stack is for ;)

**Pull request merged**